src/Shared/NamedPipeUtil.cs

```diff
@@ -25,5 +35,244 @@ internal static string GetPipeNameOrPath(string pipeName)
+
+#if !FEATURE_PIPEOPTIONS_CURRENTUSERONLY
+                if (NativeMethodsShared.IsWindows && !NativeMethodsShared.IsMono)
+                {
+                    // Verify that the owner of the pipe is us.  This prevents a security hole where a remote node has
+                    // been faked up with ACLs that would let us attach to it.  It could then issue fake build requests back to
+                    // us, potentially causing us to execute builds that do harmful or unexpected things.  The pipe owner can
+                    // only be set to the user's own SID by a normal, unprivileged process.  The conditions where a faked up
+                    // remote node could set the owner to something else would also let it change owners on other objects, so
+                    // this would be a security flaw upstream of us.
+                    ValidateRemotePipeSecurityOnWindows(nodeStream);
+                }
+#endif
+                if (handshake.HasValue)
+                {
+
+                    int[] handshakeComponents = handshake.Value.RetrieveHandshakeComponents();
+                    for (int i = 0; i < handshakeComponents.Length; i++)
+                    {
+                        CommunicationsUtilities.Trace("Writing handshake part {0} to pipe {1}", i, pipeName);
+                        nodeStream.WriteIntForHandshake(handshakeComponents[i]);
+                    }
+
+                    // This indicates that we have finished all the parts of our handshake; hopefully the endpoint has as well.
+                    nodeStream.WriteEndOfHandshakeSignal();
+
+                    CommunicationsUtilities.Trace("Reading handshake from pipe {0}", pipeName);
+
+#if NETCOREAPP2_1 || MONO
+                    nodeStream.ReadEndOfHandshakeSignal(true, timeout);
+#else
+                    nodeStream.ReadEndOfHandshakeSignal(true);
+#endif
+                }
+
+                // We got a connection.
+                CommunicationsUtilities.Trace("Successfully connected to pipe {0}...!", pipeName);
+                return nodeStream;
+            }
+            catch (Exception e) when (!ExceptionHandling.IsCriticalException(e))
+            {
+                // Can be:
+                // UnauthorizedAccessException -- Couldn't connect, might not be a node.
+                // IOException -- Couldn't connect, already in use.
+                // TimeoutException -- Couldn't connect, might not be a node.
+                // InvalidOperationException – Couldn’t connect, probably a different build
+                CommunicationsUtilities.Trace("Failed to connect to pipe {0}. {1}", pipeName, e.Message.TrimEnd());
+
+                // If we don't close any stream, we might hang up the child
+                nodeStream?.Dispose();
+            }
+
+            return null;
+        }
+
+        internal static bool ValidateHandshake(Handshake handshake, NamedPipeServerStream serverStream, int clientConnectTimeout)
```

Comment:
I was momentarily confused why you had two methods for checking the handshake right next to each other; then I realized that one was writing, and the other was reading. 👍